EventListener/ControllerArgumentOptionsListener.php

```diff
@@ -0,0 +1,138 @@
+<?php
+
+/*
+ * This file is part of the Symfony package.
+ *
+ * (c) Fabien Potencier <fabien@symfony.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Sensio\Bundle\FrameworkExtraBundle\EventListener;
+
+use Doctrine\Common\Annotations\Reader;
+use Doctrine\Common\Util\ClassUtils;
+use phpDocumentor\Reflection\DocBlockFactory;
+use phpDocumentor\Reflection\Types\Array_;
+use phpDocumentor\Reflection\Types\Object_;
+use Sensio\Bundle\FrameworkExtraBundle\Annotation\Arg;
+use Sensio\Bundle\FrameworkExtraBundle\Exception\LogicException;
+use Symfony\Component\HttpFoundation\Request;
+use Symfony\Component\HttpKernel\ControllerMetadata\ArgumentMetadata;
+use Symfony\Component\HttpKernel\ControllerMetadata\ArgumentMetadataFactory;
+use Symfony\Component\HttpKernel\Event\FilterControllerEvent;
+use Symfony\Component\HttpKernel\KernelEvents;
+use Symfony\Component\EventDispatcher\EventSubscriberInterface;
+use Symfony\Component\OptionsResolver\Exception\ExceptionInterface as OptionsResolverExceptionInterface;
+use Symfony\Component\OptionsResolver\OptionsResolver;
+
+/**
+ * Parses annotation blocks located in controller classes to extract @Arg ones.
+ *
+ * @author Fabien Potencier <fabien@symfony.com>
+ */
+class ControllerArgumentOptionsListener implements EventSubscriberInterface
+{
+    /**
+     * @var Reader
+     */
+    private $reader;
+    private $metadataFactory;
+    private $argumentOptions;
+    private $docBlockFactory;
+
+    public function __construct(Reader $reader, ArgumentMetadataFactory $metadataFactory = null, array $argumentOptions)
+    {
+        $this->reader = $reader;
+        $this->metadataFactory = $metadataFactory;
+        $this->argumentOptions = $argumentOptions;
+    }
+
+    public function onKernelController(FilterControllerEvent $event)
+    {
+        $controller = $event->getController();
+
+        if (!is_array($controller) && method_exists($controller, '__invoke')) {
+            $controller = array($controller, '__invoke');
+        }
+
+        if (!is_array($controller)) {
+            return;
+        }
+
+        $className = class_exists('Doctrine\Common\Util\ClassUtils') ? ClassUtils::getClass($controller[0]) : get_class($controller[0]);
+        $method = new \ReflectionMethod($className, $controller[1]);
+        $annotations = $this->reader->getMethodAnnotations($method);
+        $context = $className.'::'.$controller[1].'()';
+        $argumentMetadata = $this->metadataFactory->createArgumentMetadata($controller);
```

Comment:
I was just thinking this might need some internal caching so that it's only generated once. Arguments will now be generated in this method *and* the HttpKernel.

Could probably store it in an array internally at least for the current request:
https://github.com/symfony/symfony/blob/master/src/Symfony/Component/HttpKernel/ControllerMetadata/ArgumentMetadataFactory.php